src/Tolerance/Operation/Runner/RetryPromiseOperationRunner.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/*
+ * This file is part of the Tolerance package.
+ *
+ * (c) Samuel ROZE <samuel.roze@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Tolerance\Operation\Runner;
+
+use Tolerance\Operation\Exception\UnsupportedOperation;
+use Tolerance\Operation\Operation;
+use Tolerance\Operation\PromiseOperation;
+use Tolerance\Operation\RetryEvaluator\AlwaysRetryEvaluator;
+use Tolerance\Operation\RetryEvaluator\NeverRetryEvaluator;
+use Tolerance\Operation\RetryEvaluator\RetryEvaluator;
+use Tolerance\Waiter\StatefulWaiter;
+use Tolerance\Waiter\Waiter;
+use Tolerance\Waiter\WaiterException;
+
+class RetryPromiseOperationRunner implements OperationRunner
+{
+    /**
+     * @var Waiter
+     */
+    private $waitStrategy;
+    /**
+     * @var RetryEvaluator
+     */
+    private $fulfilledEvaluator;
+    /**
+     * @var RetryEvaluator
+     */
+    private $rejectedEvaluator;
+
+    /**
+     * @param Waiter $waitStrategy
+     * @param RetryEvaluator|null $fulfilledEvaluator
+     * @param RetryEvaluator|null $rejectedEvaluator
+     */
+    public function __construct(Waiter $waitStrategy, RetryEvaluator $fulfilledEvaluator = null, RetryEvaluator $rejectedEvaluator = null)
+    {
+        $this->waitStrategy = $waitStrategy;
+        $this->fulfilledEvaluator = $fulfilledEvaluator ?: new NeverRetryEvaluator();
```

Comment:
Indeed, but it is the first middleware on the stack. IMHO our middleware should be the last one to be as close as possible to the request.
And this behavior can be disabled using `http_errors` option (http://docs.guzzlephp.org/en/latest/request-options.html#http-errors), so I don't think we should rely on it.